### DISCUSSION TITLE
Large Plaintext moduli

### DISCUSSION THREAD

**Discussion by rw0x0** [tuneinsight/lattigo#515] [General]
Hi guys, what does it take to modify the library to allow large plaintext primes of up to 256 bit? Thank you!

**Comment by Pro7ech** [comment]
Before I consider providing an answer, I have the following question: do you actually need the plaintext modulus to be a prime, or can it be a composite?

I guess you do (I suspect you are trying to do ECC), but just making sure

**Reply by rw0x0** [reply]
Yes exactly, trying to do ECC, so cannot easily use something else or split it up via CRT. Thank you!

**Reply by Pro7ech** [reply]
Then it would require the following changes:

- Implementing a type for u256
- Reimplementing all algorithms that operate over the plaintext modulus to use u256 instead of u64. This will propagate down to the very low level of the library, i.e. the ring package. 

So it would require a lot lot of work.

**Reply by rw0x0** [reply]
Thanks for the answer! If I would write a custom encode/encrypt and decode/decrypt with a larger plaintext modulus, the ciphertext domain should mostly be the same (at least for textbook BGV) or am I missing something? In the first step I would just need additions and ciphertext-plaintext multiplications in the encrypted domain, so all the complicated ciphertext-ciphertext multiplication stuff can be ignored in the beginning

**Reply by Pro7ech** [reply]
Encrypt decrypt doesn't need to be modified. You'll need to adapt the encoder, and all ops in the evaluator that use the plaintext modulus at some point. The hardest one will be the basis extender. And you'll also need to implement a ring that accepts u256. Probably it's best to already factorize p-1 to speed up the generation of the ring

**Reply by Pro7ech** [reply]
You will also need to update the scale in rlwe, as it is only u128

**Reply by Pro7ech** [reply]
Do you need batching? I.e. packing multiple plaintext values. If you don't then you could hack something without having to modify too much the encoder and implement a basis reduction with big.int

**Reply by rw0x0** [reply]
thanks for the detailed answer, will definitely look at it. Regarding batching, I intend to have it at some point, but I did not even check yet, whether my prime allows for it. But you might be right, a first version without batching might be easier to implement... Thanks again!
